### PR TITLE
documentation: prplMesh to boarfarm integration

### PR DESCRIPTION
**Pull request by abelog**
## Description

This PR represents proposal of way to integrate, organise and move current `test_flows.py` functionality to [boardfarm](https://github.com/mattsm/boardfarm/) project.
As tests and devices we test are not yet covered by this project, we have to implement them on our own.

## Contains:
- Classes organisation inside of `boardfarm`
- Methods, fields
- Typical set up schematic
- General single test flow

Main goal of this PR is to gather feedback, agreed on and have a plan (and documentation!) before moving to actual implementation.

## Not covered
Particular tests implementation. Test suites organisation.

**NOTE**: This PR is work in progress. But any suggestions are highly welcome.

### CODE REVIEW

**Review by VladyslavTupikin — APPROVED** [review]

**Comment by ghost** [comment]
So, in this diagram, the OpenWrtRouter class inherits from the rax40 class.
Isn't this backwards? Or are we saying that every OpenWrtRouter is a kind of rax40?

**Comment by abelog** [comment]
It's backwards, you are right. Fixed!

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    wait_for_network()
+    get_memfree()
+    start_tinyproxy()
+    take_lock()
+    release_lock()
+    perform_curl()
+}
+
+class OpenWrtRouter {
+    string promt
```
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]:
```suggestion
    string prompt
```
?

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    IPv4Network wan_network
+    IPv4Address wan_gateway
+    prplMesh_agent [..]
+
+    kill_console_at_exit()
+    check_logs()
+    check_status()
+}
+
+class backhaul_sniffer {
```
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Although the terms fronthaul_sniffer and backhaul_sniffer are technically correct, I find them confusing. I would prefer to call them wired_sniffer and wireless_sniffer.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    string model
+
+    __init__()
+    start_monitor()
+    stop_monitor()
+    catch_message()
+    set_ssid()
+    set_bsid()
+    get_channel()
+    get_bandwith()
```
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: AFAIU, the wireless sniffer should use monitor mode and therefore it should be configured with the channel to monitor, not with ssid and bssid.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+class prplMesh_agent {
+    int port
+    IPv4Address ip
+    prplMesh_repeater[..]
```
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: This is already represented with the relationship `*--` below, so no need to repeat it here.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml [hunk 5]
```diff
@@ -0,0 +1,165 @@
+@startuml
+class LinuxDevice {
+    string tftp_dir
```
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: I think the UML file is there to convey a message, not to have a one-to-one description of the code - the code is better at doing that :-). So, I think only the members that are really interesting (i.e. will be used in our tests) should be mentioned here.

**Review by arnout — COMMENTED** [review]
I'm always a bit ambivalent about including UML files in the repo, because they are always inconsistent with reality. So I would be more inclined to put it on the wiki, since there things are expected to be inconsistent :-). That said, the whole point of having UML files is to be able to discuss the architecture, and that's easier in a PR than on the wiki.

Maybe an intermediate solution is to handle it in an issue.

Anyway, about the architecture proposal: looks OK to me in general. I think it should be merged with the UML file that I created to get an overall picture of where we are going.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 2 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Actually, difference between wireless and wired sniffer is in "listener" set up: 
- For wired we just have to call `tcpdump` and point ports + iface to listen to (probably setting proto will also make sense), just to minimize garbage amount and ease search.
- For wireless we have to additionally check monitor capability, put WLAN iface to monitor mode.

Most part of methods, like some abstract `wait_for_capability_report` will be same for both sniffers. So, we even may split it in 3 classes: 
- generic sniffer (has methods to work with `tcpdump`, which are common)
- wired sniffer, extends "generic sniffer", extends with constructor to bring wired `tcpdump`
- wireless sniffer, extends "generic sniffer" with wireless monitor set up methods + constructor

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 3 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: I wanted to use kind of "filter" (look just into frames with specific SSID of BSSID during waiting for some action)  with `tcpdump` to reduce amount of messages to search through.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 5 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Fixed in update! I have left only crucial fields and methods in boardfarm classes.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 4 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Fixed!

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 1 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Fixed! Thanks for paying attention.

**Comment by abelog** [comment]
Current state of diagram as image:
image

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml (hunk 2 above)
Comment (abelog) [reply documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: Updated UML to demonstrate proposal of split up.

**Review by arnout — APPROVED** [review]
I don't think it makes sense to discuss more at the level of a UML file, let's just start implementing this and hopefully update the UML file if it turns out a little different. For sure, there is nothing here that is conflicting with what I have in my mind.

**Pull request merged**